File: go/signing_keys_test.go
package realmid

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestSigningKeys_ListReadsKeyringAndPolicy(t *testing.T) {
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	mux.HandleFunc("/platforms/"+testRealmID+"/signing-keys", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "GET" {
			t.Errorf("method=%s", r.Method)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"keys": []any{
				map[string]any{"kid": "k2", "created_at": 200, "active_until": 900, "retire_at": 1200, "is_current": true},
				map[string]any{"kid": "k1", "created_at": 100, "active_until": 200, "retire_at": 500, "is_current": false},
			},
			"rotation": map[string]any{"mode": "auto", "interval": "1w", "next_rotation_at": 900},
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	out, err := r.SigningKeys.List(context.Background())
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if len(out.Keys) != 2 {
		t.Fatalf("keys=%d", len(out.Keys))
	}
	if !out.Keys[0].IsCurrent {
		t.Errorf("first key should be current")
	}
	if out.Rotation.Mode != "auto" || out.Rotation.Interval != "1w" || out.Rotation.NextRotationAt != 900 {
		t.Errorf("rotation=%+v", out.Rotation)
	}
}

func TestSigningKeys_RotatePostsAndReturnsKIDs(t *testing.T) {
	mux := http.NewServeMux()
	mintPlatformToken(mux)
	mux.HandleFunc("/platforms/"+testRealmID+"/signing-keys/rotate", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("method=%s", r.Method)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"kid": "k3", "retired_kids": []string{"k1"}})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	out, err := r.SigningKeys.Rotate(context.Background())
	if err != nil {
		t.Fatalf("rotate: %v", err)
	}
	if out.KID != "k3" {
		t.Errorf("kid=%q", out.KID)
	}
	if len(out.RetiredKIDs) != 1 || out.RetiredKIDs[0] != "k1" {
		t.Errorf("retired=%v", out.RetiredKIDs)
	}
}

File: ts/src/signing-keys.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";

interface Captured {
  method: string;
  url: string;
  body?: unknown;
}

function mkFetch(handler: (req: Captured) => Response): typeof fetch {
  return (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    const method = init?.method ?? "GET";
    if (url.endsWith("/auth/login")) {
      return new Response(JSON.stringify({
        status: "ok", subject_type: "platform", refresh_token: "rtok", access_token: "pt_x", expires_in: 300,
      }), { status: 200, headers: { "content-type": "application/json" } });
    }
    let body: unknown;
    if (typeof init?.body === "string" && init.body.length > 0) {
      try { body = JSON.parse(init.body); } catch { body = init.body; }
    }
    return handler({ method, url, body });
  }) as typeof fetch;
}

test("signingKeys.list: reads keyring + rotation policy", async () => {
  const fetch = mkFetch((req) => {
    assert.equal(req.method, "GET");
    assert.match(req.url, /\/platforms\/r\/signing-keys$/);
    return new Response(JSON.stringify({
      keys: [
        { kid: "k2", created_at: 200, active_until: 900, retire_at: 1200, is_current: true },
        { kid: "k1", created_at: 100, active_until: 200, retire_at: 500, is_current: false },
      ],
      rotation: { mode: "auto", interval: "1w", next_rotation_at: 900 },
    }), { status: 200, headers: { "content-type": "application/json" } });
  });
  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const out = await realm.signingKeys.list();
  assert.equal(out.keys.length, 2);
  assert.equal(out.keys[0]!.is_current, true);
  assert.equal(out.rotation.mode, "auto");
  assert.equal(out.rotation.interval, "1w");
  assert.equal(out.rotation.next_rotation_at, 900);
});

test("signingKeys.list: tolerates a bare/empty envelope", async () => {
  const fetch = mkFetch(() => new Response(JSON.stringify({}), {
    status: 200, headers: { "content-type": "application/json" },
  }));
  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const out = await realm.signingKeys.list();
  assert.deepEqual(out.keys, []);
  assert.equal(out.rotation.mode, "auto");
});

test("signingKeys.rotate: POSTs rotate and returns new/retired kids", async () => {
  const fetch = mkFetch((req) => {
    assert.equal(req.method, "POST");
    assert.match(req.url, /\/platforms\/r\/signing-keys\/rotate$/);
    return new Response(JSON.stringify({ kid: "k3", retired_kids: ["k1"] }), {
      status: 200, headers: { "content-type": "application/json" },
    });
  });
  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const out = await realm.signingKeys.rotate();
  assert.equal(out.kid, "k3");
  assert.deepEqual(out.retired_kids, ["k1"]);
});
